src/Build/BackEnd/Components/RequestBuilder/IntrinsicTasks/ItemGroupIntrinsicTask.cs

```diff
@@ -215,21 +215,18 @@ private void ExecuteAdd(ProjectItemGroupTaskItemInstance child, ItemBucket bucke
                 FileSystems.Default,
                 LoggingContext);
 
-            if (LogTaskInputs && !LoggingContext.LoggingService.OnlyLogCriticalEvents && itemsToAdd?.Count > 0)
-            {
+            // Now add the items we created to the lookup.
+            bucket.Lookup.AddNewItemsOfItemType(child.ItemType, itemsToAdd, !keepDuplicates, (itemList) => {
                 ItemGroupLoggingHelper.LogTaskParameter(
                     LoggingContext,
                     TaskParameterMessageKind.AddItem,
                     parameterName: null,
                     propertyName: null,
                     child.ItemType,
-                    itemsToAdd,
+                    itemList,
                     logItemMetadata: true,
                     child.Location);
-            }
-
-            // Now add the items we created to the lookup.
-            bucket.Lookup.AddNewItemsOfItemType(child.ItemType, itemsToAdd, !keepDuplicates); // Add in one operation for potential copy-on-write
```

Comment:
does it work if you swap this two operations - first add then log? why you pass lambda instead?